manheim_c7n_tools/dryrun_diff.py

```diff
@@ -100,6 +112,87 @@ def _find_changed_policies(self, git_dir=None, diff_against='master'):
             pnames.append(m.group(1))
         return pnames
 
+    def _make_diff_report(self, dryrun):
+        """
+        Return a HTML report breaking down the differences between the dryrun
+        (this branch) and the last run of each policy on master.
+
+        Requires a jinja template located at `./reporting-template/report.j2`
+        within the same directory where the dryrun-diff step has been run.
+
+        :param dryrun: dryrun policy resource information
+        :type dryrun: dict
+        :return: html report
+        :rtype: str
+        """
+        all_policies = list(
+            set(dryrun.keys()) | set(self._live_results.keys())
+        )
+        entries = {}
+        if not all_policies:
+            logger.info('no policies found - skipping diff report')
+            return ''
+        t_loader = FileSystemLoader(searchpath="./reporting-template/")
+        t_env = Environment(loader=t_loader)
+        t_file = "report.j2"
+        try:
+            tmpl = t_env.get_template(t_file)
+        except TemplateNotFound:
+            logger.info('unable to find a template - skipping diff report')
+            return ''
+
+        for policy in all_policies:
+            policy_splits = policy.split('-', 1)
+            if len(policy_splits) <= 1:
+                logger.warning(
+                    "skipping policy: \
+                        %s from report - it is not prefixed correctly",
+                    policy)
+                continue
+            resource_type = policy_splits[0]
```

Comment:
Assuming that the policy name begins with the resource type isn't going to work for most users, including my own organization.

I know that this will complicate things quite a bit, but alongside the ``resources.json`` files that c7n writes with the policy stats, it also writes the policy itself to a ``metadata.json`` (or ``metadata.json.gz``) file. The right way to find the resource type for each policy would be, in the methods where we get the policy stats (``_get_dryrun_results()`` for the dry-run, and ``_get_s3_results_for_region()`` which calls ``_get_latest_res_for_policy()``) we should also retrieve the metadata files. We can then get the resource type from that.

I know this module is relatively rough code that's not terribly well documented and has no unit tests... so I'll leave it up to you... if you want to try implementing that, cool. If not, I think I'm willing to merge this as-is and then I'll add in that logic myself. Just let me know.

Reply:
Yeah that sounds like a nice solution - I'd be happy to include it in the PR :)